### PR TITLE
Logto Auth Backend plugin

### PR DESCRIPTION
**Pull request by alfonsograziano**
### Description

Warning: at this stage hot reloading is not working properly, we might improve that in another PR
This is only the first implementation, mainly to create the `api-auth` package. Will change and export multiple decorations in further PRs

## Type

- [ ] **Dependency upgrade**
- [ ] **Bug fix**
- [X] **New feature**
- [ ] **Dev change**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,13 +11,15 @@
...
     "update-env": "MODE=update node scripts/init-env.mjs",
     "init:packages": "npm i",
-    "init:ds": "[ -d packages/design-system/dist ] || npm run build:ds", 
+    "init:ds": "[ -d packages/design-system/dist ] || npm run build:ds",
+    "init:api-auth": "npm run build:api-auth",
     "init:env": "node scripts/init-env.mjs",
     "init:logto": "node scripts/init-logto.mjs",
-    "stop:logto": "node scripts/stop-logto.mjs",
-    "init": "npm run init:packages && npm run init:env && npm run init:ds && npm run stop:logto && npm run init:logto",
+    "stop:logto": "echo pippo",
```
Comment (SamSalvatico) [inline package.json:21]: Maybe _goofy_ would be more international 😂

File: packages/api-auth/index.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
+import { createRemoteJWKSet, jwtVerify } from "jose";
+import { FastifyInstance, FastifyRequest, FastifyReply } from "fastify";
+import fp from "fastify-plugin";
+
+const extractBearerToken = (authHeader: string) => {
+  const [type, token] = authHeader.split(" ");
+  if (type !== "Bearer") {
+    throw new Error("Invalid Authorization header type");
```
Comment (SamSalvatico) [inline packages/api-auth/index.ts:8]: Can you please add `...type, 'Bearer' expected`

Thank you

File: package.json (hunk 1 above)
Comment (alfonsograziano) [reply package.json:21]: Uh sorry about that, I shouldn't have committed that line lol

File: apps/payments/app/[locale]/(hosted)/logto_integration/config.ts [hunk 3]
```diff
@@ -1,16 +1,57 @@
...
+};
 
-  scopes: [AuthUserScope.Organizations, AuthUserScope.OrganizationRoles],
+// All the permissions of a normal citizen
+const citizenScopes = ["payments:create:payment"];
+const paymentsPublicServantScopes = ["payments:create:providers"];
+
+export default {
+  ...baseConfig,
+  // All the availabie resources to the app
```
Comment (SamSalvatico) [inline apps/payments/app/[locale]/(hosted)/logto_integration/config.ts:27]: minor: `availabie` typo

File: packages/api-auth/index.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    async (req: FastifyRequest, rep: FastifyReply, permissions: string[]) => {
+      const authHeader = req.headers.authorization;
+      if (!authHeader) {
+        rep.status(401).send({ message: "Unauthorized" });
+        return;
+      }
+      try {
+        await checkPermissions(authHeader, opts, permissions);
+      } catch (e) {
+        console.log(e);
```
Comment (SamSalvatico) [inline packages/api-auth/index.ts:72]: should we remove this?

**Review by SamSalvatico — APPROVED** [review]
Great work guys

File: apps/payments/app/[locale]/(hosted)/logto_integration/config.ts (hunk 3 above)
Comment (alfonsograziano) [reply apps/payments/app/[locale]/(hosted)/logto_integration/config.ts]: good catch, fixed :)

File: packages/api-auth/index.ts (hunk 4 above)
Comment (alfonsograziano) [reply packages/api-auth/index.ts]: fixed!

**Pull request merged**